scene/resources/texture.cpp

```diff
@@ -1803,6 +1807,280 @@ Ref<Image> GradientTexture::get_data() const {
+			Gradient &g = **gradient;
+			// `create()` isn't available for non-uint8_t data, so fill in the data manually.
+			image->lock();
+			for (int y = 0; y < height; y++) {
+				for (int x = 0; x < width; x++) {
+					float ofs = _get_gradient_offset_at(x, y);
+					image->set_pixel(x, y, g.get_color_at_offset(ofs));
+				}
+			}
+			image->unlock();
+		} else {
+			PoolVector<uint8_t> data;
+			data.resize(width * height * 4);
+			{
+				uint8_t *wd8 = data.write().ptr();
+				Gradient &g = **gradient;
+				for (int y = 0; y < height; y++) {
+					for (int x = 0; x < width; x++) {
+						float ofs = _get_gradient_offset_at(x, y);
+						const Color &c = g.get_color_at_offset(ofs);
+
+						wd8[(x + (y * width)) * 4 + 0] = uint8_t(CLAMP(c.r * 255.0, 0, 255));
+						wd8[(x + (y * width)) * 4 + 1] = uint8_t(CLAMP(c.g * 255.0, 0, 255));
+						wd8[(x + (y * width)) * 4 + 2] = uint8_t(CLAMP(c.b * 255.0, 0, 255));
+						wd8[(x + (y * width)) * 4 + 3] = uint8_t(CLAMP(c.a * 255.0, 0, 255));
+					}
+				}
+			}
+			image->create(width, height, false, Image::FORMAT_RGBA8, data);
+		}
+	}
+
+	if (texture.is_valid()) {
+		VS::get_singleton()->free(texture);
+		texture = VS::get_singleton()->texture_create_from_image(image);
+	} else {
+		texture = VS::get_singleton()->texture_create_from_image(image);
+	}
+
+	emit_changed();
+}
+
+float GradientTexture2D::_get_gradient_offset_at(int x, int y) const {
+	if (fill_to == fill_from) {
+		return 0;
+	}
+	float ofs = 0;
+	Vector2 pos;
+	if (width > 1) {
+		pos.x = static_cast<float>(x) / (width - 1);
+	}
+	if (height > 1) {
+		pos.y = static_cast<float>(y) / (height - 1);
+	}
+	if (fill == Fill::FILL_LINEAR) {
+		Vector2 segment[2];
+		segment[0] = fill_from;
+		segment[1] = fill_to;
+		Vector2 closest = Geometry::get_closest_point_to_segment_uncapped_2d(pos, &segment[0]);
+		ofs = (closest - fill_from).length() / (fill_to - fill_from).length();
+		if ((closest - fill_from).dot(fill_to - fill_from) < 0) {
+			ofs *= -1;
+		}
+	} else if (fill == Fill::FILL_RADIAL) {
+		ofs = (pos - fill_from).length() / (fill_to - fill_from).length();
+	}
+	if (repeat == Repeat::REPEAT_NONE) {
+		ofs = CLAMP(ofs, 0.0, 1.0);
+	} else if (repeat == Repeat::REPEAT) {
+		ofs = Math::fmod(ofs, 1.0f);
+		if (ofs < 0) {
+			ofs = 1 + ofs;
+		}
+	} else if (repeat == Repeat::REPEAT_MIRROR) {
+		ofs = Math::abs(ofs);
+		ofs = Math::fmod(ofs, 2.0f);
+		if (ofs > 1.0) {
+			ofs = 2.0 - ofs;
+		}
+	}
+	return ofs;
+}
+
+void GradientTexture2D::set_width(int p_width) {
+	width = p_width;
+	_queue_update();
+}
+
+int GradientTexture2D::get_width() const {
+	return width;
+}
+
+void GradientTexture2D::set_height(int p_height) {
+	height = p_height;
+	_queue_update();
+}
+
+int GradientTexture2D::get_height() const {
+	return height;
+}
+
+void GradientTexture2D::set_flags(uint32_t p_flags) {
+	if (p_flags == flags) {
+		return;
+	}
+
+	flags = p_flags;
+	VS::get_singleton()->texture_set_flags(texture, flags);
+	_change_notify("flags");
```

Comment:
I had to add this since 3.x stores the texture flags on each texture.
I don't completely understand the purpose of this (it seems to work fine without it?):
```cpp
_change_notify("flags");
emit_changed();
```
 but it's what `ImageTexture` does too so I'm guessing it's right?

Reply:
The implementation for `set_flags()` is empty for `GradientTexture`, so I assume it's not needed for procedurally generated textures.

Reply:
I think it is necessary to be able to apply point filtering for pixel art games etc. I was just talking about those 2 lines that I was a little confused by.